### PR TITLE
700C: Use vector ops for rx_filter_coh() when possible.

### PR DESCRIPTION
**Pull request by tmiw**
Using vector ops (via clang/gcc specific syntax + memcpy/memmove) results in ~67% less time in rx_filter_coh() without affecting receive:

Screen Shot 2021-07-24 at 12 57 16 PM

This supports issue #199 (which contains numbers for rx_filter_coh() prior to this PR).

### CODE REVIEW

File: src/cohpsk.c [hunk 1]
```diff
@@ -771,7 +771,24 @@ void fdm_downconvert_coh(COMP rx_baseband[COHPSK_NC][COHPSK_M+COHPSK_M/P], int N
     }
 }
 
+/* Determine if we can use vector ops below. */
+#if __GNUC__ > 4 || \
```
Comment (tmiw) [inline src/cohpsk.c:775]: @drowe67, I think this is correct to ensure that the vector calls are used only for supporting compilers (e.g. [this](https://stackoverflow.com/questions/1771945/c-how-to-access-elements-of-vector-using-gcc-sse-vector-extension) has a comment indicating that vector subscript addressing was introduced in GCC 4.6), but we may want to double-check to be sure.

**Comment by tmiw** [comment]
UT failures:

```
The following tests FAILED:
	  4 - test_CML_ldpcut (Failed)
	 12 - test_OFDM_modem_octave_port (Failed)
	 21 - test_OFDM_modem_octave_datac0_mpp_coded (Failed)
	 32 - test_OFDM_modem_datac0_octave_burst (Failed)
	 33 - test_OFDM_modem_datac1_octave (Failed)
	 34 - test_OFDM_modem_datac3_octave (Failed)
	 85 - test_fsk_lib_4fsk_ldpc (Failed)
```

However, these don't seem related to 700C. I'll rerun from master to be sure.

**Comment by drowe67** [comment]
@tmiw 67% less time is a 3x speed up?  Well done, that's great :+1: Nice work in modifying the DSP code for vector operations.  FYI some other methods of doing this in LPCNet (coding assembler macros for SIMD).

Can you pls use the command line method in https://github.com/drowe67/codec2/pull/185 to measure performance?  This is more portable across systems and lets us test the modem directly without freedv-gui.  It's probably more important to see how it improves performance on smaller machines, like a Pi 3, we already know 700C runs fine on powerful machines.

Was there also a cmake file you need to check in with the compiler flags?  Or is it just a C level thing?

Pls don't release any freedv-gui tests builds using this PR - I want to keep testing freedv-gui against codec2/master so we can get to codec2 1.0.  I also have several codec 2 PRs waiting that I will merge after that.

**Comment by tmiw** [comment]
ARM64 test results for rx_filter_coh() (M1 Mac Mini):

|**Branch**|**30 seconds no audio (%)**|**30 seconds ve9qrp_700c.wav (%)**|
|---|---|---|
|master|25.2%|2.7%|
|ms-700c-optimization|15.1%|1.4%|

Note that [Neon is considered unsafe by GCC for 32 bit ARM](https://en.wikipedia.org/wiki/ARM_architecture#Advanced_SIMD_(Neon)) and requires a compiler flag to enable, so the same results may not apply on Linux. Fortunately GCC should do the right thing and not use Neon for the vector ops.

In addition, assuming it was indeed okay to use -funsafe-math-optimizations, it's possible that we'll need to switch the code to using float4 instead of float8. [GCC's documentation](https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html) isn't exactly clear on that, though.

**Comment by tmiw** [comment]
> Can you pls use the command line method in #185 to measure performance? This is more portable across systems and lets us test the modem directly without freedv-gui. It's probably more important to see how it improves performance on smaller machines, like a Pi 3, we already know 700C runs fine on powerful machines.

Using the method in #185 on my Intel MBP and the code in this PR:

```
Mooneers-16-MacBook-Pro:build_osx mooneer$ time ./src/freedv_rx 700C ~/devel/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m2.834s
user	0m2.532s
sys	0m0.008s
Mooneers-16-MacBook-Pro:build_osx mooneer$ time ./src/freedv_rx 700C ~/devel/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m0.303s
user	0m0.297s
sys	0m0.004s
Mooneers-16-MacBook-Pro:build_osx mooneer$ 
```

On master:

```
Mooneers-16-MacBook-Pro:build_osx mooneer$ time ./src/freedv_rx 700C ~/devel/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m5.407s
user	0m5.136s
sys	0m0.011s
Mooneers-16-MacBook-Pro:build_osx mooneer$ time ./src/freedv_rx 700C ~/devel/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m0.519s
user	0m0.512s
sys	0m0.004s
Mooneers-16-MacBook-Pro:build_osx mooneer$ 
```

> Was there also a cmake file you need to check in with the compiler flags? Or is it just a C level thing?

No cmake changes, just C code. It might be useful to have gcc/clang generate a .s file for cohpsk.c to ensure AVX, etc. are actually being used.

> Pls don't release any freedv-gui tests builds using this PR - I want to keep testing freedv-gui against codec2/master so we can get to codec2 1.0. I also have several codec 2 PRs waiting that I will merge after that.

No worries, this will be after 1.0. 👍

**Comment by tmiw** [comment]
ARM32 results (Raspberry Pi 4), this PR without -funsafe-math-optimizations:

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m17.511s
user	0m17.464s
sys	0m0.032s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m1.927s
user	0m1.893s
sys	0m0.032s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $
```

master without -funsafe-math-optimizations:

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m22.746s
user	0m22.732s
sys	0m0.012s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m2.362s
user	0m2.350s
sys	0m0.012s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ 
```

This PR with tweaked compiler flags:

```
-- Compiler Flags:  -Wall -Wno-strict-overflow -std=gnu11 -fPIC-g -DDUMP -march=armv8-a+crc+simd -mtune=cortex-a72 -O3 -funsafe-math-optimizations

...

pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m16.002s
user	0m15.991s
sys	0m0.011s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m1.791s
user	0m1.780s
sys	0m0.011s
```

master:

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m18.468s
user	0m18.447s
sys	0m0.021s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m2.027s
user	0m2.017s
sys	0m0.011s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ 
```

There may be more of an improvement on one of the 64-bit distros. I'll have to install one and test more.

**Comment by tmiw** [comment]
ARM64 results (Raspberry Pi 4B):

master, default compiler options:

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m18.130s
user	0m18.124s
sys	0m0.004s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m1.863s
user	0m1.858s
sys	0m0.001s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ 
```

This PR, default compiler options:

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m23.379s
user	0m23.363s
sys	0m0.012s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m2.311s
user	0m2.303s
sys	0m0.008s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ 
```

This PR, with overridden compiler flags (same as above):

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m23.565s
user	0m23.552s
sys	0m0.012s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m2.348s
user	0m2.343s
sys	0m0.004s
```

Not sure why the 64 bit version of Raspberry Pi OS is causing this PR to take *more* time to execute than other 64 bit ARM machines. Bugs in the distro still?

**Comment by tmiw** [comment]
I made some progress on the Raspberry Pi 4 tonight by implementing a NEON version of this PR. Results on 64 bit Raspberry Pi OS with default compiler flags:

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m14.776s
user	0m14.771s
sys	0m0.004s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m1.557s
user	0m1.549s
sys	0m0.008s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ 
```

**Comment by drowe67** [comment]
Great work @tmiw - those are significant gains.  This concept would be useful for the recent HF data work too, the acquisition routines are quiet CPU intensive there too.  It's a neat trick to do it with just some C code re-arrangement :+1:

**Comment by tmiw** [comment]
OK, here are the results I have so far:

**x86_64 (2019 Intel MBP)**

```
HT-TM05:build_osx mooneer$ time ./src/freedv_rx 700C ../../wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m3.258s
user	0m2.945s
sys	0m0.005s
HT-TM05:build_osx mooneer$ time ./src/freedv_rx 700C ../../wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m0.345s
user	0m0.339s
sys	0m0.004s
HT-TM05:build_osx mooneer$ 
```

**ARM64 Raspberry Pi 4B**

```
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700d.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m13.346s
user	0m13.341s
sys	0m0.005s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $ time ./src/freedv_rx 700C ~/freedv-gui/wav/ve9qrp_700c.wav /dev/null
frames decoded: 1404  output speech samples: 898560

real	0m1.388s
user	0m1.384s
sys	0m0.005s
pi@raspberrypi:~/freedv-gui/codec2/build_linux $
```

Intel got a bit worse because I switched from a 256 bit vector to a 128 bit one (to better optimize for ARM) but still better than what it was before. If we really wanted to we could possibly #ifdef some Intel specific code that still used 256 bit vectors. 👍

**Comment by drowe67** [comment]
Nice work @tmiw - significant improvements across several platforms :+1:

**Comment by tmiw** [comment]
@drowe67, thanks! I'm wondering if there's anything that can be done algorithm-wise as well given that rx_filter_coh() is still using most of the CPU when executing 700C RX. There might not be, though, and if not the performance on the systems people have been using should still be good.

**Comment by drowe67** [comment]
Hi @tmiw, I presume you mean the acquisition CPU load?  Yes, that can be pretty heavy :thinking: It's also a "delicate" part of demodulator design - I tend to spend more time on acquisition than any other part, and it's easy to break (this happened with mods for a Flex radio port of this modem a few years ago).

We could consider some algorithm changes, for example:
1. Look at ways to re-arrange the processing steps.  In acquisition, we are running the core demod at a bunch of different frequency offsets, obtaining the demodulated symbols, and seeing which one gives us the best correlation with the pilots.  There might be another way to do this.
1. Look into FFT based implementation of the filter convolution.  Although that's usually only efficient for really big FIR filters.
1. Shorten the filters we apply to each carrier (less taps) for acquisition, currently `COHPSK_NT` is 5 symbols, we could try 3.  This would involve Octave and C changes, and we may need to add some ctests to make sure the acquisition is still solid.

**Comment by tmiw** [comment]
@drowe67, I also played with using AVX and inline assembly on Intel for rx_filter_coh(). That seems to work, but playing ve9qrp_700c.wav in freedv-gui somehow results in a lower SNR (~6dB) vs the current PR (~20dB). I'm not sure why that's the case, unfortunately.

OTOH, time ./src/freedv_rx dropped down to something like 2.3 seconds with AVX, so it may still be an option in conjunction with the rest of the PR if we figure out why the SNR's lower.

**Comment by drowe67** [comment]
@tmiw are you happy with this PR to be merged?

**Comment by tmiw** [comment]
> @tmiw are you happy with this PR to be merged?

I think so. It still would be nice to use AVX but without knowing more about why SNR drops and/or having it be a configurable option (vs. always doing it when able), that probably shouldn't be included now.

Fortunately NEON doesn't seem to have nearly the same effect on SNR (if any at all) so it can probably remain compiled in on ARM.

**Pull request merged**